pkg/operator/deploymentcontroller/deployment_controller.go

```diff
@@ -389,3 +402,95 @@ func hasFinishedProgressing(deployment *appsv1.Deployment) bool {
+// when the deployment has actually failed or when pods disappear after successful deployment.
+func isDeploymentAvailableDuringRollout(deployment *appsv1.Deployment) bool {
+	// First, check for hard failures
+	for _, cond := range deployment.Status.Conditions {
+		// ProgressDeadlineExceeded means the deployment has failed
+		if cond.Type == appsv1.DeploymentProgressing &&
+			cond.Status == corev1.ConditionFalse &&
+			cond.Reason == "ProgressDeadlineExceeded" {
+			return false
+		}
+		// ReplicaFailure means pods are failing to start
+		if cond.Type == appsv1.DeploymentReplicaFailure && cond.Status == corev1.ConditionTrue {
+			return false
+		}
+	}
+
+	// Check if deployment is actively being updated (spec change being rolled out)
+	// This is the primary indicator of a "normal upgrade" in progress
+	if deployment.Generation != deployment.Status.ObservedGeneration {
+		// Spec has changed, deployment controller is working on rolling it out
+		// Per API contract, remain Available during this normal upgrade
+		return true
+	}
+
+	// Check if we're in an active rollout (not just missing pods after successful deployment)
+	var isProgressing bool
+	var isRollingOut bool
+	var hasFinishedRollout bool
+
+	for _, cond := range deployment.Status.Conditions {
+		if cond.Type == appsv1.DeploymentProgressing {
+			if cond.Status == corev1.ConditionTrue {
+				isProgressing = true
+				// ReplicaSetUpdated means we're actively rolling out new pods
+				if cond.Reason == "ReplicaSetUpdated" {
+					isRollingOut = true
+				}
+			}
+			// NewReplicaSetAvailable means rollout completed successfully
+			if cond.Status == corev1.ConditionTrue && cond.Reason == "NewReplicaSetAvailable" {
+				hasFinishedRollout = true
+			}
+		}
+	}
+
+	// If we're actively rolling out, remain Available per API contract
+	if isRollingOut {
+		return true
+	}
+
+	// If rollout has finished successfully (NewReplicaSetAvailable) but pods are now missing,
+	// this is NOT a normal upgrade - it's an operational issue (node failure, etc.)
+	// In this case, we should report Available=False
+	if hasFinishedRollout {
+		return false
+	}
+
+	// If we have updated replicas being created, we're in a rollout
+	var expectedReplicas int32
+	if deployment.Spec.Replicas != nil {
+		expectedReplicas = *deployment.Spec.Replicas
+	}
+
+	if deployment.Status.UpdatedReplicas > 0 && deployment.Status.UpdatedReplicas <= expectedReplicas {
+		// We have updated replicas being created, this is a normal rollout
+		return true
+	}
+
+	// If progressing but no updated replicas yet, we're in early stages of rollout
+	if isProgressing {
+		return true
+	}
+
+	// Special case: brand new deployment with no status conditions yet
+	// This happens during initial deployment before Kubernetes has had a chance to update status
+	if len(deployment.Status.Conditions) == 0 && deployment.Status.ObservedGeneration == 0 {
+		// Deployment just created, is progressing normally
+		return true
+	}
```

Comment:
This is just wrong.

1. The code is not reachable
2. The status should be `Available = False` during installation and `Available = true` during upgrade. This code has no clue about that.